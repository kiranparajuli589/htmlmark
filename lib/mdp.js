import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
