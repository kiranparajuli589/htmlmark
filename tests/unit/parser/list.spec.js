import {MDP} from "../../mdp.js"


const mdp = new MDP()


describe("List Parsing", () => {
	it("should tokenize a valid ordered list", () => {
		const lines = [
			"1. item **1**",
			"1. item [link](link-url)",
			"1. item 3 `code item`"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should tokenize a valid un-ordered list", () => {
		const lines = [
			"- item **1**",
			"- item [link](link-url)",
			"- item 3 `code item`"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should detect list indent", () => {
		const lines = [
			"  - item **1**",
			"  - item [link](link-url)",
			"  - item 3 `code item`"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("indent should break the list", () => {
		const lines = [
			"- item **1**",
			"- item [link](link-url)",
			"    - item 3 `code item`",
			"        - item 4",
			"    - item 3 `code item`"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should tokenize list combination", () => {
		const lines = [
			"- one",
			"- two",
			"1. one",
			"2. two",
			"- [ ] c empty",
			"- [x] c checked",
			"1. [ ] c empty",
			"1. [x] c checked"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse the indented codeblock inside", () => {
		const lines = [
			"To put a code block within a list item, the code block needs",
			"to be indented *twice* -- 8 spaces or two tabs:",
			"*   A list item with a code block:",
			"",
			"        <code goes here>",
			"        <code goes here>",
			"        <code goes here>",
			"",
			"### Code Blocks"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should allow empty list item", () => {
		const lines = [
			"- one",
			"-",
			"- two"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it.each([
		{ lines: [
			"- one",
			"-",
			"- two"
		] },
		{ lines: [
			"-",
			"- one",
			"- two"
		] },
		{ lines: [
			"- one",
			"- two",
			"-"
		] }
	])("should parse a list with an empty list item", ({ lines }) => {
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse a spaced list", () => {
		const lines = [
			"",
			"- here we go with a list item",
			"",
			"- another list item is here",
			"",
			"- next list item is on the fly"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
})
