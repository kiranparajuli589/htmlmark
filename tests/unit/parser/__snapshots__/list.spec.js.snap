// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Parsing indent should break the list 1`] = `
"<ul>
<li><p>item <strong>1</strong></p>
</li>
<li><p>item <a href=\\"link-url\\">link</a></p>
<ul>
<li><p>item 3 <code>code item</code></p>
<ul>
<li><p>item 4</p>
</li>
</ul>
</li>
<li><p>item 3 <code>code item</code></p>
</li>
</ul>
</li>
</ul>
"
`;

exports[`List Parsing should allow empty list item 1`] = `
"<ul>
<li><p>one</p>
</li>
<li></li>
<li><p>two</p>
</li>
</ul>
"
`;

exports[`List Parsing should detect list indent 1`] = `
"<ul>
<li><p>item <strong>1</strong></p>
</li>
<li><p>item <a href=\\"link-url\\">link</a></p>
</li>
<li><p>item 3 <code>code item</code></p>
</li>
</ul>
"
`;

exports[`List Parsing should parse a list with an empty list item 1`] = `
"<ul>
<li><p>one</p>
</li>
<li></li>
<li><p>two</p>
</li>
</ul>
"
`;

exports[`List Parsing should parse a list with an empty list item 2`] = `
"<ul>
<li></li>
<li><p>one</p>
</li>
<li><p>two</p>
</li>
</ul>
"
`;

exports[`List Parsing should parse a list with an empty list item 3`] = `
"<ul>
<li><p>one</p>
</li>
<li><p>two</p>
</li>
<li></li>
</ul>
"
`;

exports[`List Parsing should parse a spaced list 1`] = `
"<ul>
<li><p>here we go with a list item</p>
</li>
<li><p>another list item is here</p>
</li>
<li><p>next list item is on the fly</p>
</li>
</ul>
"
`;

exports[`List Parsing should parse the indented codeblock inside 1`] = `
"<p>To put a code block within a list item, the code block needs to be indented <em>twice</em> -- 8 spaces or two tabs:</p>
<ul>
<li><p>A list item with a code block:</p>
<pre><code>&lt;code goes here&gt;
&lt;code goes here&gt;
&lt;code goes here&gt;</code></pre>
</li>
</ul>
<h3>Code Blocks</h3>
"
`;

exports[`List Parsing should tokenize a valid ordered list 1`] = `
"<ol>
<li><p>item <strong>1</strong></p>
</li>
<li><p>item <a href=\\"link-url\\">link</a></p>
</li>
<li><p>item 3 <code>code item</code></p>
</li>
</ol>
"
`;

exports[`List Parsing should tokenize a valid un-ordered list 1`] = `
"<ul>
<li><p>item <strong>1</strong></p>
</li>
<li><p>item <a href=\\"link-url\\">link</a></p>
</li>
<li><p>item 3 <code>code item</code></p>
</li>
</ul>
"
`;

exports[`List Parsing should tokenize list combination 1`] = `
"<ul>
<li><p>one</p>
</li>
<li><p>two</p>
</li>
</ul>
<ol>
<li><p>one</p>
</li>
<li><p>two</p>
</li>
</ol>
<ul>
<li><input type='checkbox'><p>c empty</p>
</li>
<li><input type='checkbox' checked><p>c checked</p>
</li>
</ul>
<ol>
<li><input type='checkbox'><p>c empty</p>
</li>
<li><input type='checkbox' checked><p>c checked</p>
</li>
</ol>
"
`;
